src/machine/machinectl.c

```diff
@@ -124,20 +135,30 @@ static void clean_machine_info(MachineInfo *machines, size_t n_machines) {
         if (!machines || n_machines == 0)
                 return;
 
-        for (i = 0; i < n_machines; i++)
+        for (i = 0; i < n_machines; i++) {
                 free(machines[i].os);
+                free(machines[i].version_id);
+        }
         free(machines);
 }
 
-static int get_os_name(sd_bus *bus, const char *name, char **out) {
+static int get_os_release_property(sd_bus *bus, const char *name, const char *query, ...) {
         _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
-        const char *k, *v, *os;
-        char *str;
+        _cleanup_free_ const char **query_res = NULL;
+        const char *k, *v, *iter;
+        size_t count = 0, awaited_args = 0;
+        va_list ap;
         int r;
 
         assert(bus);
         assert(name);
-        assert(out);
+        assert(query);
+
+        NULSTR_FOREACH(iter, query)
+                awaited_args++;
+        query_res = new0(const char *, awaited_args);
```

Comment:
given that this is bounded I think newa0() is your friend hier (i.e. allocate this on the stack)